src/phoenix/experimental/evals/models/openai.py

```diff
@@ -210,6 +233,25 @@ def _build_messages(
     def verbose_generation_info(self) -> str:
         return f"OpenAI invocation parameters: {self.public_invocation_params}"
 
+    async def _async_generate(self, prompt: str, **kwargs: Any) -> str:
+        invoke_params = self.invocation_params
+        messages = self._build_messages(prompt, kwargs.get("instruction"))
+        if functions := kwargs.get("functions"):
+            invoke_params["functions"] = functions
+        if function_call := kwargs.get("function_call"):
+            invoke_params["function_call"] = function_call
+        response = await self._async_generate_with_retry(
+            messages=messages,
+            **invoke_params,
+        )
+        choice = response["choices"][0]
+        if self._model_uses_legacy_completion_api:
+            return str(choice["text"])
+        message = choice["message"]
+        if function_call := message.get("function_call"):
+            return str(function_call.get("arguments") or "")
+        return str(message["content"])
```

Comment:
this is the same as below? maybe we DRY it up a bit?

Reply:
Absolutely, I left the synchronous codepath here as a precaution. I didn't rule out also making a "synchronous executor" which just recovers the existing loop behavior so we can allow users to choose what submission strategy they want to use.

If we think this isn't necessary though, we can remove the sync codepath altogether too

Reply:
sg, let's leave it for now